This line is causing problems

```
freq = index.freq if index.freq else index.inferred_freq
```

so is this one:

```
freq = getattr(endog.index, 'inferred_freq', None)
```

Also seeing this error:

```
    if not isinstance(freq, (str, None)):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   TypeError: isinstance() arg 2 must be a type, a tuple of types, or a union
```

--- a/statsmodels/compat/pandas.py
+++ b/statsmodels/compat/pandas.py
@@ -93,7 +93,8 @@ def infer_freq(index) -> str | None:
         # in older versions the option is not available and a str is returned
         freq = pd.infer_freq(index)
 
-    if not isinstance(freq, (str, None)):
+    # new pandas versions retuns BaseOffset
+    if not isinstance(freq, str) and freq is not None:
         return freq.freqstr
     return freq
 

--- a/statsmodels/compat/pandas.py
+++ b/statsmodels/compat/pandas.py
@@ -1,4 +1,5 @@
 from __future__ import annotations
+from contextlib import contextmanager
 
 from typing import TYPE_CHECKING, Any, Callable, Mapping, Optional, TypeVar
 
@@ -83,14 +84,21 @@ assert_index_equal = testing.assert_index_equal
 assert_series_equal = testing.assert_series_equal
 
 
-def infer_freq(index) -> str | None:
+@contextmanager
+def _infer_freq_returns_offset():
 
     # pandas 3.1 changes the return value of infer_freq
     try:
         with pd.option_context("future.infer_freq_returns_offset", True):
-            freq = pd.infer_freq(index)
+            yield
     except pd.errors.OptionError:
         # in older versions the option is not available and a str is returned
+        yield
+
+def infer_freq(index) -> str | None:
+
+    # pandas 3.1 changes the return value of infer_freq
+    with _infer_freq_returns_offset():
         freq = pd.infer_freq(index)
 
     # new pandas versions retuns BaseOffset

--- a/statsmodels/tsa/base/tsa_model.py
+++ b/statsmodels/tsa/base/tsa_model.py
@@ -4,6 +4,7 @@ from statsmodels.compat.pandas import (
     is_float_index,
     is_int_index,
     is_numeric_dtype,
+    _infer_freq_returns_offset,
 )
 
 import numbers
@@ -568,7 +569,8 @@ class TimeSeriesModel(base.LikelihoodModel):
             if isinstance(index, (DatetimeIndex, PeriodIndex)):
                 # If no frequency, try to get an inferred frequency
                 if freq is None and index.freq is None:
-                    freq = index.inferred_freq
+                    with _infer_freq_returns_offset():
+                        freq = index.inferred_freq
                     # If we got an inferred frequncy, alert the user
                     if freq is not None:
                         inferred_freq = True

--- a/statsmodels/tsa/deterministic.py
+++ b/statsmodels/tsa/deterministic.py
@@ -24,6 +24,7 @@ from statsmodels.tools.validation import (
     string_like,
 )
 from statsmodels.tsa.tsatools import freq_to_period
+from statsmodels.compat.pandas import _infer_freq_returns_offset
 
 if TYPE_CHECKING:
     from collections.abc import Hashable, Sequence
@@ -392,7 +393,8 @@ class Seasonality(DeterministicTerm):
         if isinstance(index, pd.PeriodIndex):
             freq = index.freq
         elif isinstance(index, pd.DatetimeIndex):
-            freq = index.freq if index.freq else index.inferred_freq
+            with _infer_freq_returns_offset():
+                freq = index.freq if index.freq else index.inferred_freq
         else:
             raise TypeError("index must be a DatetimeIndex or PeriodIndex")
         if freq is None:
@@ -1438,7 +1440,8 @@ differences can be extended when producing out-of-sample forecasts.
             self._index_freq = self._index.freq
             self._extendable = True
         elif isinstance(self._index, pd.DatetimeIndex):
-            self._index_freq = self._index.freq or self._index.inferred_freq
+            with _infer_freq_returns_offset():
+                self._index_freq = self._index.freq or self._index.inferred_freq
             self._extendable = self._index_freq is not None
         elif isinstance(self._index, pd.RangeIndex):
             self._extendable = True

--- a/statsmodels/tsa/filters/_utils.py
+++ b/statsmodels/tsa/filters/_utils.py
@@ -2,7 +2,7 @@ from functools import wraps
 
 from statsmodels.tools.data import _is_using_pandas
 from statsmodels.tsa.tsatools import freq_to_period
-
+from statsmodels.compat.pandas import _infer_freq_returns_offset
 
 def _get_pandas_wrapper(X, trim_head=None, trim_tail=None, names=None):
     index = X.index
@@ -83,7 +83,8 @@ def pandas_wrapper_freq(func, trim_head=None, trim_tail=None,
         wrapper_func = _get_pandas_wrapper(X, trim_head, trim_tail,
                                            columns)
         index = X.index
-        freq = index.inferred_freq
+        with _infer_freq_returns_offset():
+            freq = index.inferred_freq
         kwargs.update({freq_kw : freq_to_period(freq)})
         ret = func(X, *args, **kwargs)
         ret = wrapper_func(ret)

--- a/statsmodels/tsa/stl/_stl.pyx
+++ b/statsmodels/tsa/stl/_stl.pyx
@@ -89,6 +89,7 @@ from libc.math cimport NAN, fabs, isnan, sqrt
 from statsmodels.tools.validation import array_like
 from statsmodels.tsa.seasonal._seasonal import DecomposeResult
 from statsmodels.tsa.tsatools import freq_to_period
+from statsmodels.compat.pandas import _infer_freq_returns_offset
 
 
 def _is_pos_int(x, odd):
@@ -219,7 +220,8 @@ cdef class STL(object):
         if period is None:
             freq = None
             if isinstance(endog, (pd.Series, pd.DataFrame)):
-                freq = getattr(endog.index, 'inferred_freq', None)
+                with _infer_freq_returns_offset():
+                    freq = getattr(endog.index, 'inferred_freq', None)
             if freq is None:
                 raise ValueError('Unable to determine period from endog')
             period = freq_to_period(freq)

--- a/statsmodels/tsa/vector_ar/tests/test_var.py
+++ b/statsmodels/tsa/vector_ar/tests/test_var.py
@@ -2,7 +2,7 @@
 Test VAR Model
 """
 
-from statsmodels.compat.pandas import QUARTER_END, assert_index_equal
+from statsmodels.compat.pandas import QUARTER_END, assert_index_equal, _infer_freq_returns_offset
 from statsmodels.compat.python import lrange
 
 from io import BytesIO
@@ -953,7 +953,8 @@ def test_correct_nobs():
     mdata = mdata[["realgdp", "realcons", "realinv"]]
     mdata.index = pd.DatetimeIndex(quarterly)
     data = np.log(mdata).diff().dropna()
-    data.index.freq = data.index.inferred_freq
+    with _infer_freq_returns_offset():
+        data.index.freq = data.index.inferred_freq
     data_exog = pd.DataFrame(index=data.index)
     rs = np.random.RandomState(238971)
     data_exog["exovar1"] = rs.normal(size=data_exog.shape[0])

